modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -100,21 +101,50 @@ def __init__(
 
         self._final_checkpoint_path.mkdir()  # exist_ok == False, this directory should not exist before
 
-        self._status_query_queue = status_query_queue
-        self._status_response_queue = status_response_queue
+        self._status_query_queue_training = status_query_queue_training
+        self._status_response_queue_training = status_response_queue_training
+
+        self._status_query_queue_downsampling = status_query_queue_downsampling
+        self._status_response_queue_downsampling = status_response_queue_downsampling
 
         self._num_samples = 0
 
         self._metadata_collector = MetadataCollector(training_info.pipeline_id, training_info.trigger_id)
 
         self.selector_stub = self.connect_to_selector(training_info.selector_address)
-        self._downsampling_enabled, strategy_name, params_from_selector = self.get_selection_strategy()
-        self._weighted_opt = False
+        self.selector_address = training_info.selector_address
+        downsampling_enabled, strategy_name, downsampler_config = self.get_selection_strategy()
 
-        if self._downsampling_enabled:
+        if downsampling_enabled:
             self._criterion_nored = criterion_func(**training_info.criterion_dict, reduction="none")
-            self._downsampler = self.instantiate_downsampler(strategy_name, params_from_selector, self._criterion_nored)
+            self._downsampler = self.instantiate_downsampler(strategy_name, downsampler_config, self._criterion_nored)
+
+            assert "sample_then_batch" in downsampler_config
+            if downsampler_config["sample_then_batch"]:
+                self._downsampling_mode = DownsamplingMode.SAMPLE_THEN_BATCH
+                assert "downsampling_period" in downsampler_config
+                self._downsampling_period = downsampler_config["downsampling_period"]
+            else:
+                self._downsampling_mode = DownsamplingMode.BATCH_THEN_SAMPLE
             self._weighted_opt = True
+        else:
+            self._downsampling_mode = DownsamplingMode.DISABLED
+            self._weighted_opt = False
```

Comment:
Is that correct? What about weights from the selector? Right now we only set `self._weighted_opt` to `True` if we downsample as I understand it.

Could we add a unit test testing all of these behaviors and asserting it is doing the right thing?

Reply:
Yes, the weights were not managed but intentionally. I have now added two parameters within the train function for this purpose.

The first one in `retrieve_weights_from_dataloader` that is True if the dataloader returns the weights (aka dataloader._uses_weights is true). The second one is `weighted_optimization` which is true if we get the weights from the dataloader (`retrieve_weights_from_dataloader = True`) or if we compute the weights in the training loop (so, right now, just if we use BTS).

The function is quite simple

```
def weights_handling(self) -> Tuple[bool, bool]:
    # whether the dataloader returns the weights
    retrieve_weights_from_dataloader = self._train_dataloader.dataset.retrieve_weights_from_dataloader()
    # we want to use weighted optimization if we get weights from the dataloader or if we compute them in the
    # training loop (BATCH_THEN_SAMPLE downsampling mode)
    weighted_optimization = (
        retrieve_weights_from_dataloader or self._downsampling_mode == DownsamplingMode.BATCH_THEN_SAMPLE
    )

    return retrieve_weights_from_dataloader, weighted_optimization
```

And then the first parameter is used to unpack the batch, so 
```
if retrieve_weights_from_dataloader:
    weights = batch[3]
```
And the second one to apply weighted_sgd:
```
if weighted_optimization:
    # weighted gradient descent
    assert weights is not None
    loss = torch.dot(self._criterion_nored(output, target), weights / weights.sum())
else:
    loss = self._criterion(output, target)
```

The only thing I do not like 100% is that this function (weights_handling) is called within the training loop (thus once per batch). 
The problem is that if we use the selector as the key source, it requires that the GRPC connection has been made and this can only happen after the first __iter__ (otherwise there are problems creating multiple processes). I honestly don't see a better alternative and in any case it is reading a dataset parameter, so it has no impact on performance

Reply:
I had to change `retrieve_weights_from_dataloader = self._train_dataloader.dataset.retrieve_weights_from_dataloader()` since the grpc connection is made only within the worker dataloader and making it in the main thread breaks the seralization (can't be pickled). Now, instead of calling a function, we just check if we have weights (aka len(batch) == 4). It's not the most elegant solution, but the length of the batch is 4 only if the strategy uses weights. So now the function is

```
def weights_handling(self, batch_len: int) -> Tuple[bool, bool]:
    # whether the dataloader returned the weights.
    retrieve_weights_from_dataloader = batch_len == 4 # key, sample, label, weight

    # we want to use weighted optimization if we get weights from the dataloader or if we compute them in the
    # training loop (BATCH_THEN_SAMPLE downsampling mode)
    weighted_optimization = (
        retrieve_weights_from_dataloader or self._downsampling_mode == DownsamplingMode.BATCH_THEN_SAMPLE
    )

    return retrieve_weights_from_dataloader, weighted_optimization

```

Reply:
Thank you, I think the solution which is checking for weights == 4 makes sense in our current setup. The two bools make everything more clear now. Thank you!